### PR TITLE
Implementing create document.

### PR DESCRIPTION
**Pull request by fernamig**

### CODE REVIEW

File: OnBaseDocsApi/Controllers/OnBaseDocsController.cs [hunk 1]
```diff
@@ -57,7 +60,188 @@ public IHttpActionResult GetFile(int id)
...
+                    DocumentType = toStaging ? Global.Config.StagingDocType : docAttr.DocumentType,
+                    Comment = docAttr.Comment,
+                    IndexKey = docAttr.IndexKey,
+                    Keywords = docAttr.Keywords,
+                    Ext = docExt,
+                    Stream = docStream,
+                    ToStaging = toStaging,
+                };
+                var result = CreateDocument(app, createAttr, out var doc);
+                if (result != null)
```
Comment (tsoliangwu0130) [inline OnBaseDocsApi/Controllers/OnBaseDocsController.cs:166]: Correct me if I'm wrong, looks like `CreateDocument` will return `null` only if everything looks good? If so, I think `result` here is misleading; we should probalbly name it `errors` or something else.

Comment (fernamig) [reply OnBaseDocsApi/Controllers/OnBaseDocsController.cs:166]: Good point.

File: OnBaseDocsApi/Controllers/OnBaseDocsController.cs [hunk 2]
```diff
@@ -57,7 +60,188 @@ public IHttpActionResult GetFile(int id)
...
+            });
+        }
+
+        IHttpActionResult CreateDocument(Application app, DocumentCreateAttributes attr, out Document doc)
+        {
+            doc = null;
+
+            var docType = app.Core.DocumentTypes.Find(attr.DocumentType);
+            if (docType == null)
+                return InternalErrorResult($"The DocumentType '{attr.DocumentType}' could not be found.");
```
[the remark this answers is not in the document]
Comment (fernamig) [reply OnBaseDocsApi/Controllers/OnBaseDocsController.cs]: Ok.

File: OnBaseDocsApi/Controllers/OnBaseDocsController.cs [hunk 3]
```diff
@@ -57,7 +60,188 @@ public IHttpActionResult GetFile(int id)
...
+            }
+
+            if ((docStream == null) || string.IsNullOrWhiteSpace(docExt))
+                return BadRequestResult("The required parameter 'file' is missing.");
+            if (docAttr == null)
+                return BadRequestResult("The required parameter 'attributes' is missing.");
+            if (string.IsNullOrWhiteSpace(docAttr.IndexKey))
+                return BadRequestResult("The required parameter 'IndexKey' is empty.");
+            if (string.IsNullOrWhiteSpace(docAttr.DocumentType))
+                return BadRequestResult("The required parameter 'DocumentType' is empty.");
```
[the remark this answers is not in the document]
Comment (fernamig) [reply OnBaseDocsApi/Controllers/OnBaseDocsController.cs]: OK.

File: OnBaseDocsApi/Controllers/OnBaseDocsController.cs [hunk 4]
```diff
@@ -57,7 +61,200 @@ public IHttpActionResult GetFile(int id)
...
+                }
+            }
+
+            var pageData = app.Core.Storage.CreatePageData(attr.Stream, attr.Ext);
+            if (pageData == null)
+                InternalErrorResult($"Unable to create page data for '{attr.Ext}'.");
+
+            doc = app.Core.Storage.StoreNewDocument(pageData, props);
+            if (doc == null)
+                return InternalErrorResult($"Unable to create document.");
```
Comment (tsoliangwu0130) [inline OnBaseDocsApi/Controllers/OnBaseDocsController.cs:238]: I think we don't need template string here?

File: OnBaseDocsApi/Controllers/OnBaseDocsController.cs [hunk 5]
```diff
@@ -57,7 +61,200 @@ public IHttpActionResult GetFile(int id)
...
+            try
+            {
+                var provider = await Request.Content.ReadAsMultipartAsync(
+                    new MultipartMemoryStreamProvider());
+
+                foreach (HttpContent content in provider.Contents)
+                {
+                    // The Content-Disposition header is required.
+                    if (content.Headers.ContentDisposition == null)
+                        return BadRequestResult($"A Content-Disposition header is required.");
```
Comment (tsoliangwu0130) [inline OnBaseDocsApi/Controllers/OnBaseDocsController.cs:90]: Same, I think we can use normal string here.

File: OnBaseDocsApi/Controllers/OnBaseDocsController.cs [hunk 6]
```diff
@@ -57,7 +61,200 @@ public IHttpActionResult GetFile(int id)
...
+            {
+                foreach (var kw in attr.Keywords)
+                {
+                    props.AddKeyword(kw.Name, kw.Value);
+                }
+            }
+
+            var pageData = app.Core.Storage.CreatePageData(attr.Stream, attr.Ext);
+            if (pageData == null)
+                InternalErrorResult($"Unable to create page data for '{attr.Ext}'.");
```
Comment (tsoliangwu0130) [inline OnBaseDocsApi/Controllers/OnBaseDocsController.cs:234]: I guess this should be bad request as well?

File: OnBaseDocsApi/Controllers/OnBaseDocsController.cs (hunk 5 above)
Comment (fernamig) [reply OnBaseDocsApi/Controllers/OnBaseDocsController.cs]: Agree.

File: OnBaseDocsApi/Controllers/OnBaseDocsController.cs [hunk 7]
```diff
@@ -57,7 +61,200 @@ public IHttpActionResult GetFile(int id)
...
+            if (docType == null)
+                return BadRequestResult($"The DocumentType '{attr.DocumentType}' could not be found.");
+
+            var fileType = app.Core.FileTypes.Find(attr.Ext);
+            if (fileType == null)
+                return BadRequestResult($"The FileType '{attr.Ext}' could not be found.");
+
+            var props = app.Core.Storage.CreateStoreNewDocumentProperties(docType, fileType);
+            if (props == null)
+                return InternalErrorResult($"Unable to create document properties for '{attr.DocumentType}' and '{attr.Ext}'.");
```
[the remark this answers is not in the document]
Comment (fernamig) [reply OnBaseDocsApi/Controllers/OnBaseDocsController.cs]: Yeah, it should.

File: OnBaseDocsApi/Controllers/OnBaseDocsController.cs (hunk 6 above)
Comment (fernamig) [reply OnBaseDocsApi/Controllers/OnBaseDocsController.cs]: Yup.

File: OnBaseDocsApi/Controllers/OnBaseDocsController.cs (hunk 4 above)
Comment (fernamig) [reply OnBaseDocsApi/Controllers/OnBaseDocsController.cs]: Agree.

**Pull request merged**